### DISCUSSION TITLE
How to add more source files for the external scanner

### DISCUSSION THREAD

**Discussion by jpt13653903** [tree-sitter/tree-sitter#3398] [Q&A] [answered]
I'm trying to add more than just `scanner.c` as source files to the external scanner.  So far I've updated:

`bindings/go/binding.go:4`:
```go
// #include "../../src/parser.c"
// #include "../../src/scanner.c"
// #include "../../src/TokenTree.c"
```
`bindings/rust/build.rs:10`
```rust
let parser_path = src_dir.join("parser.c");
c_config.file(&parser_path);
println!("cargo:rerun-if-changed={}", parser_path.to_str().unwrap());

let scanner_path = src_dir.join("scanner.c");
c_config.file(&scanner_path);
println!("cargo:rerun-if-changed={}", scanner_path.to_str().unwrap());

let token_tree_path = src_dir.join("TokenTree.c");
c_config.file(&token_tree_path);
println!("cargo:rerun-if-changed={}", token_tree_path.to_str().unwrap());
```

`binding.gyp`:13
```gyp
"src/parser.c",
"src/scanner.c",
"src/TokenTree.c",
```

`Makefile`:32 seems to auto-add all `*.c` files, so I left it alone

`Package.swift:36`
```swift
sources: [
    "src/parser.c",
    "src/scanner.c",
    "src/TokenTree.c",
],
```

`setup.py`:36
```python
sources=[
    "bindings/python/tree_sitter_vhdl/binding.c",
    "src/parser.c",
    "src/scanner.c",
    "src/TokenTree.c",
],
```

-------------

With all these updates, `tree-sitter parse` still returns with `undefined reference` errors for all the symbols in `TokenTree.c`.  I really hope that I'm doing something stupid -- my temporary solution of adding `#include "TokenTree.c"` to `scanner.c` is not optimal.

My build environment is:

- Windows 11 version 10.0.22631.3593
- Git Bash terminal (git version 2.42.0)
- npm version 10.8.0
- T-Dragon MinGW (i.e. gcc) version 10.3.0

**Comment by amaanq** [comment, marked as answer]
yeah the only way you can get this to work is to include the C files - adding complexity is not something we want to do by allowing multiple C source/header files. In practice, it's just not needed at all and if it is for templating/boilerplate/large static objects you can just directly include it.

**Reply by jpt13653903** [reply]
Thanks - I can work with that.  Good to know the limitations of the framework.

I'll turn everything into headers, because I’m scared that the Makefile `*.c` will result in duplicate declarations in some cases.

**Reply by amaanq** [reply]
yeah a header is a good idea for sure, here's an example of where we make use of one:

https://github.com/tree-sitter/tree-sitter-html/blob/master/src/tag.h
